### PR TITLE
Improve "infrastructure" crate wide

### LINKED ISSUE #75: Document features
I saw at least a lack of `doc(cfg(...))` on `std::error::Error` impl there's likely more.

### PR DESCRIPTION
**Pull request by tcharding**
Improve various things around the crate (infrastructure for want of a better word)

**Please note:** Includes bump of MSRV to 1.48.0 because arrayvec was failing with 1.41 after improving CI.

EDIT: Oooooh, now I get it, arrayvec feature breaks the MSRV, regardless of if its 1.41 or 1.48 - I forgot that. I"m totally sick of rebasing this so I'm leaving the MSRV bump in there.

- CI
- github actions
- linting
- docs
- githooks

Patch 7 `ef35d89 Enable build of docs with doc_auto_cfg` fixes #75

### CODE REVIEW

**Comment by tcharding** [comment]
a

**Comment by tcharding** [comment]
Please do not merge until https://github.com/rust-bitcoin/rust-bech32/pull/45 is done so @Ericson2314 does not have to rebase on top of this.

**Comment by tcharding** [comment]
Checked over the whole patch series, made minor improvements removing an unrelated change from one patch. 

Added an additional patch to do a clippy fix.

File: githooks/pre-commit [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  git config hooks.allownonascii true
+EOF
+	exit 1
+fi
+
+# If there are whitespace errors, print the offending file names and fail.
+git diff-index --check --cached $against -- || exit 1
+
+# Check that code lints cleanly.
+cargo clippy --all-features -- -D warnings || exit 1
```
Comment (clarkmoody) [inline githooks/pre-commit:49]: This is a cool addition. Confirmed working after opting in to hooks.

Comment (tcharding) [reply githooks/pre-commit:49]: Yeah the git hooks are super useful, it pains me to wait each time I commit so I often use `-n` which ignores the hooks i.e., `git commit -a -m 'WIP: dirty commit' -n`

**Comment by apoelstra** [comment]
I guess the docsrs thing should be changed now to use `doc_auto_cfg` or whatever that was.

**Comment by apoelstra** [comment]
ooo TIL about `git config --local core.hooksPath githooks/` I may actually start using your hooks now that I don't have to manually copy them into .git/hooks..

**Review by apoelstra — APPROVED** [review]
ACK c691a83cf4f9476537d135b6e5c4f4af1ff72e65

**Pull request merged**